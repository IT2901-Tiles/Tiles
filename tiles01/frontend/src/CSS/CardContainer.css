/* MIDDLE CONTAINER*/

.CardContainer{
    background-color:#FBF4D7;
    color: black;
    margin: 3vw 3vw;
    height: 100%;
    position: relative;
    padding-bottom: 2vw;
}

.CardContainerTitle {
    transform-origin: 0% 0%;
    transform: rotate(-90deg) ;
    height: 2vh;
    background-color: #FBF4D7;
    font-family: 'Roboto', sans-serif;
    letter-spacing: 0.1ex;
    position: absolute;
    font-size: 3vw;
    left: -2%;
    bottom: 35%;
    padding: 12px 12px;
    z-index: 1;
}

/* Makes sure that the CardContainer does not overlap TopComponent */
@media only screen and (max-width: 700px) {
    .CardContainer {
        margin-top: 4vw;
      }
}
