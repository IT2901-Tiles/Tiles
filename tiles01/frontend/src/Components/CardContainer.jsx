import React, { useState } from 'react';
import Card from "./Card"

//imports all of the images
function importAll(paths) {
    return paths.keys().map(paths);
}

//snippet from https://stackoverflow.com/questions/38332094/how-can-i-mock-webpacks-require-context-in-jest
if (typeof require.context === 'undefined') {
	const fs = require('fs');
	const path = require('path');

	require.context = (base = '.', scanSubDirectories = false, regularExpression = /\.js$/) => {
	    const files = {};
	    function readDirectory(directory) {
	        fs.readdirSync(directory).forEach((file) => {
	            const fullPath = path.resolve(directory, file);

	            if (fs.statSync(fullPath).isDirectory()) {
	                if (scanSubDirectories) readDirectory(fullPath);
	    
                    return;
	            }
	            
                if (!regularExpression.test(fullPath)) return;
	            
                files[fullPath] = true;
	        });
	    }

	    readDirectory(path.resolve(__dirname, base));
	    
        function Module(file) {
	        return require(file);
	    }
	    
        Module.keys = () => Object.keys(files);
	    
        return Module;
	    };
	}

//paths to where the different image sets are
var cardPaths = {
    "things": require.context("./../images/things", false, /\.(png|jpe?g|svg)$/),
    "trigger": require.context("./../images/triggers", false, /\.(png|jpe?g|svg)$/),
    "feedback": require.context("./../images/feedbacks", false, /\.(png|jpe?g|svg)$/)
}

function CardContainer(props) {
    const listOfImages = importAll(cardPaths[props.category]);
    const [activeCard, setActiveCard] = useState(null);

    //called when a card in the container is clicked, sets the card as active 
    //newValue represents the card component that is selected
    function handleCardClick(newValue) {
        props.onSelect(newValue)
        setActiveCard(newValue.id)
    }
    //gets card name from path	
	function getName(path) {
        let index;
        try{
            index = path.default.indexOf(".")
        }catch{
            return null;
        }
        return path.default.slice(14, index)
    }
    
    //map function maps the files imported in listOfImages to Card components
    return (
        <div>
            {listOfImages.map(
                (image, index) => <Card active={activeCard} name={getName(image)} onClick={handleCardClick} id={index} key={index} src={image.default} />
            )}
        </div>
    );
}

export default CardContainer;