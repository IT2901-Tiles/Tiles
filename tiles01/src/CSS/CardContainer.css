/* MIDDLE CONTAINER*/
/* 
    Må se hvordan jeg skal løse de ulike middle componentene
*/

.CardContainer{
    background-color:#FBF4D7;
    color: black;
    margin: 1vh 3vw;
    height: 70vh;
    position:relative;
}

.CardContainerTitle {
    transform-origin: 0% 0%;
    transform: rotate(-90deg) ;
    height: 2vh;
    background-color: #FBF4D7;
    font-family: 'Roboto', sans-serif;
    letter-spacing: 0.1ex;
    position: absolute;
    font-size: 5vh;
    left: -2%;
    bottom: 35%;
    padding: 1vh 1vw;
    z-index: 1;
}

