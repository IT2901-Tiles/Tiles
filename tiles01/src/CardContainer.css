/* MIDDLE CONTAINER*/
/* 
    Må se hvordan jeg skal løse de ulike middle componentene
*/

.CardContainer{
    background-color:#FBF4D7;
    color: black;
    margin: 1% 3%;
    height:32rem;
}

.CardContainerTitle {
    transform: rotate(-90deg);
    background-color: #FBF4D7;
    /*color: #68C2C4;*/
    font-family: 'Roboto', sans-serif;
    letter-spacing: 0.1ex;
    position: fixed;
    font-size: 2rem;
    padding: 0.5% 1.5%;
    z-index: 1;
    left: -3%;
    bottom: 50%;
}
