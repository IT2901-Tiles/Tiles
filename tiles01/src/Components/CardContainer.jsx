import React, { useState } from 'react';
import Card from "./Card"

//imports all of the images
function importAll(r) {
    console.log(r)
    console.log(r.keys())
    console.log(r.keys().map(r))
    return r.keys().map(r);
  }

//paths tp where the different image sets are
var cardPaths={
    "thing":require.context("./../images/things", false, /\.(png|jpe?g|svg)$/),
    "trigger":require.context("./../images/triggers", false, /\.(png|jpe?g|svg)$/),
    "feedback":require.context("./../images/feedbacks", false, /\.(png|jpe?g|svg)$/)
}
//props:
//category: the category of the cardcontainer, used to find the cards
//onSelect: function to pass selected card to CategoryPage,
//variables:
//image: the image file
//index: the index of the image file in listOfImages

function CardContainer(props) {
    const listOfImages=importAll(cardPaths[props.category]);
    const [activeCard,setActiveCard] = useState(null);

    //called when a card in the container is clicked, sets the card as active 
    function handleCardClick(newValue){
        console.log(newValue)
        props.onSelect(newValue)
        setActiveCard(newValue.id)
    }
    //gets card name from path
    function getName(path){
        console.log(path.default)
        let index=path.default.indexOf(".")
        return path.default.slice(14,index)
    }


    //map function maps the files imported in listOfImages to Card components
    return (
        <div>
        
        {listOfImages.map(
            
            (image,index) =>  <Card active={activeCard} name={getName(image)} onClick={handleCardClick} id={index} key={index} src={image.default}  />
        )}
        </div>
    );
}

export default CardContainer;