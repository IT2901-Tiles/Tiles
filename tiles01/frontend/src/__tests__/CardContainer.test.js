import {configure, shallow} from "enzyme"	
import Adapter from '@wojtekmaj/enzyme-adapter-react-17';
import {cleanup} from '@testing-library/react';
import CardContainer from '../Components/CardContainer';
import renderer from 'react-test-renderer';
	
afterEach(() => {
	cleanup();
});

configure({
	adapter: new Adapter()
});

describe('CardContainer', () => {
	const mockFunction = jest.fn()
	const initialProps = {
	    category: "trigger",
	    onSelect: mockFunction,
	}
	const wrapper = shallow(<CardContainer {...initialProps}/>)
	it('renders correctly', () => {
	    const tree = renderer.create(<CardContainer {...initialProps}/>).toJSON();
	    expect(tree).toMatchSnapshot();
	});
	it('trigger category contains 3 cards', () => {
	    expect(wrapper.find('Card')).toHaveLength(3);
	});
	it('changes number of cards when category changes', () => {
	    const initialProps = {
	        category: "feedback",
	        onSelect: mockFunction,
	    }
	    const container = shallow(<CardContainer {...initialProps} />);
	    expect(container.find('Card')).toHaveLength(2);
	});
}); 
