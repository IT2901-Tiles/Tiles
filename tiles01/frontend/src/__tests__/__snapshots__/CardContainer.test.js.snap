// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CardContainer renders correctly 1`] = `
<div>
  <img
    alt="A card"
    className="card"
    id={0}
    name={null}
    onClick={[Function]}
  />
  <img
    alt="A card"
    className="card"
    id={1}
    name={null}
    onClick={[Function]}
  />
  <img
    alt="A card"
    className="card"
    id={2}
    name={null}
    onClick={[Function]}
  />
</div>
`;
